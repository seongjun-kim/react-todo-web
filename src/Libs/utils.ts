import React from 'react';
import { render } from '@testing-library/react';
import { BrowserRouter } from 'react-router-dom';

export const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, '', route);
  return render(ui, { wrapper: BrowserRouter });
};
