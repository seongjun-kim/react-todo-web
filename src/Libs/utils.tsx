import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter, BrowserRouter } from 'react-router-dom';
import { ToDoListProvider } from 'Contexts';
import App from 'App';

/**
 * [renderWithRouter]
 * useLocation과 같은 라우터 관련 훅을 포함한 단일 컴포넌트 테스트를 위한 렌더링 함수
 * 컴포넌트 테스트를 위한 함수이므로 Components 폴더 아래 테스트에 활용
 */
export const renderWithRouter = (ui: React.ReactElement, { route = '/' } = {}) => {
  window.history.pushState({}, '', route);
  return render(ui, { wrapper: BrowserRouter });
};

/**
 * [renderWithMemoryRouter]
 * 실제 앱에서의 구동 방식 테스트를 위한 렌더링 함수
 * 페이지 전환 테스트를 위한 함수이므로 Pages 폴더 아래 테스트에 활용
 *
 *  - initialEntries를 window.history 처럼 활용 가능
 *  - TestComponent를 활용한 앱 내 추가 컴포넌트 적용 테스트 가능
 *    - 주로 useLocation().pathname을 파악하는 용도로 활용
 */
export const renderWithMemoryRouter = (
  { initialIndex = 0, initialEntries = ['/'] } = {},
  TestComponent?: JSX.Element | JSX.Element[],
) => {
  return render(
    <MemoryRouter initialEntries={initialEntries} initialIndex={initialIndex}>
      {TestComponent}
      <ToDoListProvider>
        <App />
      </ToDoListProvider>
    </MemoryRouter>,
  );
};
