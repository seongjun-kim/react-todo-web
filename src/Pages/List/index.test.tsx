import React from 'react';
import { useLocation } from 'react-router-dom';
import { fireEvent, screen } from '@testing-library/react';
import 'jest-styled-components';

import { renderWithMemoryRouter } from 'Libs/utils';

describe('<List />', () => {
  it('renders component correctly', () => {
    const sampleData = ['ToDo 1', 'ToDo 2', 'ToDo 3'];
    localStorage.setItem('ToDoList', JSON.stringify(sampleData));

    const { container } = renderWithMemoryRouter();

    const toDoItem1 = screen.getByText(sampleData[0]);
    expect(toDoItem1).toBeInTheDocument();
    expect(toDoItem1.getAttribute('href')).toBe('/detail/0');

    const toDoItem2 = screen.getByText(sampleData[1]);
    expect(toDoItem2).toBeInTheDocument();
    expect(toDoItem2.getAttribute('href')).toBe('/detail/1');

    const toDoItem3 = screen.getByText(sampleData[2]);
    expect(toDoItem3).toBeInTheDocument();
    expect(toDoItem3.getAttribute('href')).toBe('/detail/2');

    expect(screen.getAllByText('Delete').length).toBe(3);

    const AddButton = screen.getByText('+');
    expect(AddButton).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });

  it('deletes todo item', () => {
    const sampleData = ['ToDo 1', 'ToDo 2', 'ToDo 3'];
    localStorage.setItem('ToDoList', JSON.stringify(sampleData));

    renderWithMemoryRouter();

    const toDoItem = screen.getByText(sampleData[1]);
    expect(toDoItem).toBeInTheDocument();
    fireEvent.click(toDoItem.nextElementSibling as HTMLElement); // click delete button of the toDoItem
    expect(toDoItem).not.toBeInTheDocument();
    expect(JSON.parse(localStorage.getItem('ToDoList') as string)).not.toContain([sampleData[1]]);
  });

  it('moves to detail page', () => {
    const sampleData = ['ToDo 1', 'ToDo 2', 'ToDo 3'];
    localStorage.setItem('ToDoList', JSON.stringify(sampleData));

    const TestComponent = (): JSX.Element => {
      const { pathname } = useLocation();
      return <div>{pathname}</div>;
    };
    renderWithMemoryRouter({}, <TestComponent />);

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const toDoItem = screen.getByText(sampleData[2]);
    expect(toDoItem).toBeInTheDocument();
    expect(toDoItem.getAttribute('href')).toBe('/detail/2');
    fireEvent.click(toDoItem);

    expect(url.textContent).toBe('/detail/2');
  });

  it('moves to add page', () => {
    const TestComponent = (): JSX.Element => {
      const { pathname } = useLocation();
      return <div>{pathname}</div>;
    };
    renderWithMemoryRouter({}, <TestComponent />);

    const url = screen.getByText('/');
    expect(url).toBeInTheDocument();

    const addButton = screen.getByText('+');
    expect(addButton).toBeInTheDocument();
    expect(addButton.getAttribute('href')).toBe('/add');
    fireEvent.click(addButton);

    expect(url.textContent).toBe('/add');
  });
});
