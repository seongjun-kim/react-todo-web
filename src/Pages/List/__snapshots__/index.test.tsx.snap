// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<List /> renders component correctly 1`] = `
.c7 {
  text-align: center;
  background-color: #ff1744;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c7:hover {
  background-color: #f01440;
}

.c7:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c8 {
  color: #fff;
  font-size: 16px;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border-bottom: 1px solid #bdbdbd;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 10px;
  padding: 10px;
}

.c6 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  font-size: 16px;
  margin-right: 20px;
  color: black;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4 {
  min-width: 350px;
  height: 400px;
  overflow-y: scroll;
  border: 1px solid #BDBDBD;
  margin-bottom: 20px;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background-color: #1e40ff;
}

.c2 {
  padding: 20px;
  color: #fff;
  font-size: 20px;
  font-weight: 600;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  background-color: #fff;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  padding: 20px;
  border-radius: 8px;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  position: relative;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 {
  font-size: 20px;
  color: #fff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 60px;
  height: 60px;
  border-radius: 30px;
  cursor: pointer;
  position: absolute;
  bottom: -30px;
  background-color: #304ffe;
  box-shadow: 5px 5px 10px rgba(0,0,0,0.2);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c9:hover {
  background-color: #1e40ff;
}

.c9:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c0 {
  min-height: 100vh;
  display: -webkit-box;
  display: -webkit-flexx;
  display: -ms-flexxbox;
  display: flexx;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div>
  <div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        Todo List
      </div>
    </div>
    <div
      class="c3"
    >
      <div
        class="c4"
        data-testid="toDoList"
      >
        <div
          class="c5"
        >
          <a
            class="c6"
            href="/detail/0"
          >
            ToDo 1
          </a>
          <div
            class="c7"
          >
            <div
              class="c8"
            >
              Delete
            </div>
          </div>
        </div>
        <div
          class="c5"
        >
          <a
            class="c6"
            href="/detail/1"
          >
            ToDo 2
          </a>
          <div
            class="c7"
          >
            <div
              class="c8"
            >
              Delete
            </div>
          </div>
        </div>
        <div
          class="c5"
        >
          <a
            class="c6"
            href="/detail/2"
          >
            ToDo 3
          </a>
          <div
            class="c7"
          >
            <div
              class="c8"
            >
              Delete
            </div>
          </div>
        </div>
      </div>
      <a
        class="c9"
        href="/add"
      >
        +
      </a>
    </div>
  </div>
</div>
`;
